commit 94d6db3afeca306b7f2ab4209ea772575d22104d (refs/changes/24/1424/2)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2017-02-03T11:36:46-08:00 (2 years, 8 months ago)
    
    joyent/sdc-docker#110 Docker ps filtering is not correct
    Reviewed by: Trent Mick <trentm@gmail.com>

diff --git a/lib/backends/sdc/containers.js b/lib/backends/sdc/containers.js
index f0d04aa..e026125 100644
--- a/lib/backends/sdc/containers.js
+++ b/lib/backends/sdc/containers.js
@@ -2568,6 +2568,7 @@ function getContainers(opts, callback) {
     assert.object(opts, 'opts');
     assert.object(opts.account, 'opts.account');
     assert.uuid(opts.account.uuid, 'opts.account.uuid');
+    assert.optionalBool(opts.all, 'opts.all');
     assert.object(opts.app, 'opts.app');
     assert.object(opts.app.config, 'opts.app.config');
     assert.object(opts.app.config.fwapi, 'opts.app.config.fwapi');
@@ -2593,6 +2594,16 @@ function getContainers(opts, callback) {
             return;
         }
         log.debug({filters: filters}, 'getContainers: filters');
+
+        // Issue joyent/sdc-docker#110.
+        // To be able to filter on status, we must ensure we are getting all the
+        // containers, otherwise a filter against 'exited' status would not
+        // work. This may seem a little weird (especially when there already is
+        // a flag to say if one wants all containers), but that's the way docker
+        // works.
+        if (filters.hasOwnProperty('status')) {
+            opts.all = true;
+        }
     }
 
     vasync.pipeline({arg: {}, funcs: [
diff --git a/lib/endpoints/containers.js b/lib/endpoints/containers.js
index 84ee714..43206cd 100644
--- a/lib/endpoints/containers.js
+++ b/lib/endpoints/containers.js
@@ -88,7 +88,6 @@ function containerList(req, res, next) {
     options.all = common.boolFromQueryParam(req.query.all);
     // Note: options.all is implied when using these query params.
     if (req.query.hasOwnProperty('limit')
-        || req.query.hasOwnProperty('filters')
         || req.query.hasOwnProperty('before')
         || req.query.hasOwnProperty('since'))
     {
diff --git a/test/integration/cli-filters.test.js b/test/integration/cli-filters.test.js
index 4a99f6c..2702035 100644
--- a/test/integration/cli-filters.test.js
+++ b/test/integration/cli-filters.test.js
@@ -108,6 +108,27 @@ test('container filters', function (tt) {
         + ' --filter label=fishing=fun'
         + ' --filter label=todd=cool',
         [containerName2]);
+
+    tt.test('stop container 1', function (t) {
+        cli.stop(t, {args: containerName1}, function (err) {
+            t.ifErr(err, 'docker stop ' + containerName1);
+            t.end();
+        });
+    });
+
+    // Filtering against running/stopped containers.
+    checkContainerFiltering(tt, '--filter name=' + CONTAINER_PREFIX,
+        [containerName2]);
+
+    // Filtering against running/stopped containers.
+    checkContainerFiltering(tt, '--filter name=' + CONTAINER_PREFIX
+        + ' --filter status=exited',
+        [containerName1]);
+
+    // Filtering against running/stopped containers.
+    checkContainerFiltering(tt, '--filter name=' + CONTAINER_PREFIX
+        + ' --filter status=running',
+        [containerName2]);
 });
 
 
